test/test_framework_events.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import annotations
+
+import typing
+
+import pytest
+
+import ops
+from ops import testing
+
+
+def collect_filtered(
+    *event_type: type[ops.EventBase], meta: dict[str, typing.Any] | None = None
+) -> tuple[ops.EventBase, ...]:
+    ctx = testing.Context(ops.CharmBase, meta={'name': 'ben', **(meta or {})})
+    with ctx(ctx.on.update_status(), testing.State()) as mgr:
+        return tuple(e.event_type for e in mgr.charm.on.events(*event_type).values())  # type: ignore
+
+
+def test_filtering_null():
+    base_events = set(
+        subc
+        for subc in ops.HookEvent.__subclasses__()
+        # exclude non-leaves
+        if not len(subc.__subclasses__()) > 1
+    ).union(
+        # only lifecycle event that is automatically added to charm.on
+        {ops.CollectStatusEvent}
+    )
+    secret_events = set(subc for subc in ops.SecretEvent.__subclasses__())
+    assert set(collect_filtered()) == base_events.union(secret_events)
```

Comment:
This test is rather hard to read.

Can we have test case from the API user point of view?

Something that would look like:

```py
class MyCharm(...):
    def __init__(self, ...):
        super().__init__(...)
        for k, v in self.on.event(clever-args-here):
            ...
```

Then run it through Scenario, issue events and monitor some side-effect, like call count on the callback or what not.

Or maybe like this:

```
class MyCharm(...):
    def __init__(self, ...):
        super().__init__(...)
        assert self.on.events(clever-args-here) == {
            "foo": BoundEvent(self, ops.Foo),
            ...
        }
```